### DISCUSSION TITLE
When does writeContract's "wait" callback fire?

### DISCUSSION THREAD

**Discussion by cilphex** [wevm/wagmi#1541] [Q&A] [answered]
[writeContract](https://wagmi.sh/core/actions/writeContract) returns a `wait` function, presumably so that you may tell when the tx that was successfully sent to be processed has been confirmed.

My question is when and how often does this callback get called? Is it fired once, when it has 0 confirmations, then again for 1, and then continuously?

Is it only called after the first confirmation?

It's difficult to test locally because Ganache processes transactions in new blocks instantly, so I'm trying to test and account for the case of the tx taking some time to actually confirm.

**Comment by jxom** [comment, marked as answer]
The `wait` function gets called when the transaction has been processed / mined on a block (ie. 1 confirmation). [Read more](https://docs.ethers.org/v5/api/providers/types/#providers-TransactionResponse).

**Comment by cilphex** [comment]
Thank you. It would be nice if this description could be referenced in the wagmi docs somewhere, like in the wagmi/core `writeContract` [return value description](https://wagmi.sh/core/actions/writeContract#return-value).

**Reply by tmm** [reply]
Also, worth noting that [`useWaitForTransaction`](https://wagmi.sh/react/hooks/useWaitForTransaction)/[`waitForTransaction`](https://wagmi.sh/core/actions/waitForTransaction) are preferred to `wait`, since `wait` is an ethers API wagmi does not control.

**Reply by cilphex** [reply]
Noted, thank you. In the docs for `waitForTransaction`, it would also be nice to have a link to what a `TransactionReceipt` consists of.
